### PR TITLE
iiab-footprint [iiab-diagnostics to speed up remote support of networks especially?]

### PR DESCRIPTION
**Pull request by georgejhunt**
Responds to #1575 
Creates a more automated means of gathering together network settings, and communicating them back to developers.

### CODE REVIEW

**Comment by holta** [comment]
@mikkokotila & @jvonau can you review this PR from @georgejhunt?  Towards implementing:

#1575 More automated reporting of networking/similar issues

File: scripts/footprint_features.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+contents of following directories:
+
+1. /etc/network/interfaces.d (and interfaces)
+2. /etc/sysconfig/network-scripts/if-cfg*
+3. /etc/NetworkManager/system-connections
+4. /etc/systemd/network/
+
+#### Suggested Usage ####
+1. Use a FAT formatted USB stick (FAT format is recognized by both Windows and Mac) to move /etc/iiab/footprint/<footprint.tgz> to a machine that has internet access and a browser.
+2. Upload the footprint to  https://filebin.ca, record the returned URL, and email that URL, along with a description of the symptoms, to xsce-devel@googlegroups.com.
```
Comment (holta) [inline scripts/footprint_features.md]: @georgejhunt email sent to xsce-devel@googlegroups.com (Google Groups) seems to disappear rather than notifying any of us (even mailing list administrators!) regarding these emails from non-subscribers :-(

Possibly Groups.io is far better (this is basically a clone of Group Groups mailing lists, with better community functionality...and a business model that's just as unclear?)

Conclusion: recommending folks send email to bugs (AT) iiab.io would seem to be far better for now &mdash; these emails are received and triaged by @floydianslips & I for now.

File: scripts/footprint_features.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+contents of following directories:
+
+1. /etc/network/interfaces.d (and interfaces)
+2. /etc/sysconfig/network-scripts/if-cfg*
+3. /etc/NetworkManager/system-connections
+4. /etc/systemd/network/
+
+#### Suggested Usage ####
+1. Use a FAT formatted USB stick (FAT format is recognized by both Windows and Mac) to move /etc/iiab/footprint/<footprint.tgz> to a machine that has internet access and a browser.
```
Comment (holta) [inline scripts/footprint_features.md]: Should we recommend `apt install pastebinit` and then something like `pastebinit /etc/iiab/footprint/footprint.tgz` to greatly streamline this for many people?

**Comment by georgejhunt** [comment]
When I tried pastebinit, I got an error message indicating something about
utf8, and I concluded that the end points only wanted to deal with ascii.
Targz format is really binary.

**Comment by georgejhunt** [comment]
Reading more on pastebinit, there is a set of helpers, pbput, and pbget,
which encodes, and compresses. I think that might be sufficient.

**Comment by georgejhunt** [comment]
Yes pastebinit, with the two helpers, works fine! I added the contents of
/etc/rpi-issue, which gives the date of the image, and whether lite or
desktop to  http://paste.ubuntu.com/p/xjNBbrKBFx/.
The used pbget with that url. It put the unencoded file into
/tmp/etc/iiab/footprint/  which seemed fine.

**Comment by holta** [comment]
Earlier this week @tim-moody wrote that something like `iiab-diagnostics` would be a more meaningful name.  I tend to agree.

**Comment by holta** [comment]
It seems likely we should merge this during Thursday's 10AM EDT community call (http://minutes.iiab.io).

We call always revise it later, as nec!

**Comment by holta** [comment]
@georgejhunt I may have overlooked something, but can you make sure these 2 critical files are *both* included?

- /etc/iiab/local_vars.yml
- /etc/iiab/openvpn_handle

CLARIF: the 2 files above often get out of sync, so it's critical that *both* be recorded!

**Comment by georgejhunt** [comment]
Thanks, I missed that

File: scripts/iiab-diagnostics [hunk 3]
```diff
@@ -0,0 +1,155 @@
+#!/bin/bash
+# create a diagnostic footprint of network, and OS settings
+
+read -p "\n\nPlease identiy yourself (<8 characters, no spaces)" who
```
Comment (holta) [inline scripts/iiab-diagnostics]:
```suggestion
read -p "\n\nPlease provide a name or nickname: (8 characters or less, no spaces) " who
```

File: scripts/iiab-diagnostics [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+if [ -z "$who" ]; then
+   who="noname"
+fi
+
+read -p "\n\nIf this is a request for help, what email addr for response" email
```
Comment (holta) [inline scripts/iiab-diagnostics]:
```suggestion
read -p "\n\nIf this is a request for help, what is your email address: " email
```

File: scripts/iiab-diagnostics [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+echo
+echo "=========================================================="
+echo ip addr
+ip addr
+echo
+echo "=========================================================="
+echo Checking for information about raspberry pi base image
+if [ -f /etc/rpi-issue ];then
+cat /etc/rpi-issue
+echo "stage 2 = lite; stage 3 = desktop"
```
Comment (holta) [inline scripts/iiab-diagnostics]:
```suggestion
echo "stage2 = lite; stage5 = desktop; SEE github.com/RPi-Distro/pi-gen#stage-anatomy"
```

File: scripts/iiab-diagnostics [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+echo "=========================================================="
+echo "brctl show"
+brctl show
+echo
+echo "=========================================================="
+echo "/etc/resolv.conf"
+cat /etc/resolv.conf
+echo
+echo "=========================================================="
+echo "cat  iiab_ini_file "
```
Comment (holta) [inline scripts/iiab-diagnostics]:
```suggestion
echo "/etc/iiab/iiab.ini"
```

**Comment by holta** [comment]
@georgejhunt the output of these 2 commands is also necessary:

- `systemctl status dnsmasq`
- `journalctl -u dnsnasq`

REASON: different kinds of dnsmasq failures occur constantly on new installations, for a myriad different reasons of varying hardware, varying distros, varying network topologies, varying upstream/downstream patches — the list goes on.

Solving these existential quagmires 10X more efficiently *(with more data, during that very first call/contact)* is frankly THE only way forward for us — to establish progressively more trusted+understandable baseline infrastructure — stepping it up for implementers in all countries, one month at a time.


QUESTION: how can we make each of these reports viewable on a smartphone (not just .tgz) for strong support volunteers like Jerry, who are unable to use a full computer, while at work and on the road?

**Comment by georgejhunt** [comment]
https://apkpure.com/androzip%E2%84%A2-free-file-manager/com.agilesoftresource

**Comment by holta** [comment]
> https://apkpure.com/androzip%E2%84%A2-free-file-manager/com.agilesoftresource

That's one option.  But I remain quite concerned this will structurally lower Internet-in-a-Box's global volunteerism rate, unintentionally preventing informal civic communications, in a world that's no longer centered around `vi` as in times past.

WHAT WE NEED: short URL's that can be passed around on universally global platforms like Facebook, WhatsApp & Skype, so bug reports are not accidentally silo'd and bureaucratized.

An organic and potentially exponentially increasing community interest in solving the bugs that are most centrally affecting hundreds/thousands around the planet is a target we should not shy away from~

**Comment by holta** [comment]
FYI collecting the [bug] reporter's email address might need to be done thru a separate channel.

For both legal reasons (we don't want IIAB to be illegal in Europe under GDPR).

And also to make an organic IIAB Global Community of grassroots field reports broadly possible — in a truly participatory way — rather than accidentally walling ourselves off from globally popular / contemporary platforms.

File: scripts/iiab-diagnostics [hunk 7]
```diff
@@ -98,6 +98,18 @@ echo "/usr/bin/iiab-gen-iptable"
 cat /usr/bin/iiab-gen-iptables
 echo
 echo "=========================================================="
+echo "/.iiab-image"
+cat /.iiab-image
```
Comment (holta) [inline scripts/iiab-diagnostics:102]: `cat /.iiab-image` command will fail when `/.iiab-image` does not exist, as is often the case.

Wrap it in a bash `if` statement presumably?

**Comment by georgejhunt** [comment]
a lot of other cat'd files do not exist.  The set -e is not active

**Comment by holta** [comment]
#1744 with #1702 have illustrated a few problems we still need to resolve:

- the emerging `iiab-diagnostics` command definitely also needs to include `/etc/iiab/config_vars.yml` which shows Admin Console settings / configs
- the `pastebinit` command tragically fails on illegal characters as shown in #1741 and #1744.

  - Background: vi-readable output is fine for a some hackers but human-readable output within pastebin URL's is needed as that is rapidly actionable by all (e.g. phone users in the developing-developer-world etc, assorted folks who assist in troubleshooting/debugging).  So possibly we will need to refine (or modify) our use of the `pastebinit` command, so that it stops choking so often, when it inevitably encounters odd characters...

**Comment by holta** [comment]
So many issues like #1744 are a great reminder that the `iiab-diagnostics` command should work even (especially!) at times when DNS is *not* working on the IIAB in question.

Specifically: whatever commands (similar to `pastebinit`) that we use in the end, should be tested to be resilient and work as well as possible even when DNS is not working.

As even in 2019, DNS inevitably keeps failing for countless different reasons, in the low-budget / low-bandwidth situations where we work especially... !

Related: #1687 #1747 #1749 (etc)

**Comment by holta** [comment]
@georgejhunt as discussed during the call today (http://minutes.iiab.io) you might consider something like `more /etc/iiab/iiab.env /etc/iiab/iiab.ini /etc/iiab/local_vars.yml | pastebinit` to place a reasonably complete header above each file, as follows?
```
::::::::::::::
/etc/iiab/iiab.env
::::::::::::::
bla
bla
::::::::::::::
/etc/iiab/iiab.ini
::::::::::::::
bla
bla
bla
::::::::::::::
/etc/iiab/local_vars.yml
::::::::::::::
bla
bla
bla
bla
```
i.e. the `more` command is one decent way to put a complete path+filename above each file, without having to program that up yourself.

**Comment by georgejhunt** [comment]
That's pretty slick. I had always used less. But more is more.

**Comment by holta** [comment]
@georgejhunt if `pastebinit` proves too brittle (i.e. if it keeps choking on certain 'illegal' characters, and I've seen other errors issued by pastebinit too) are there other pastebin tools like https://gist.github.com that (might!) be more resilient?

FWIW `gist.github.com` supports anonymous and intentionally-privacy-conscious snippet posting (apparently) better than most advertising-driven pastebin sites.

But I'm not familiar enough with it to know for sure:

https://www.labnol.org/internet/github-gist-tutorial/28499/
https://stackoverflow.com/questions/6767518/what-is-the-difference-between-github-and-gist

Or, if in the end the `pastebinit` command is unavoidable, e.g. if it's well-supported across many Linux distros, perhaps we just need to run a sed/awk filter to clean out the illegal characters, before piping to whatever-pastebin-tool appears most future-proof ?

(Such sed/awk scripts must be very standard if nec!)

**Comment by georgejhunt** [comment]
Please examine this paste for what is wanted:
http://paste.ubuntu.com/p/n6HRBb8vMP/

**Comment by holta** [comment]
> Please examine this paste for what is wanted:
> http://paste.ubuntu.com/p/n6HRBb8vMP/

Consider these 7 ideas/recommendations to evolve on that?

1) Please include strong-visual-line-breaks before and after each header, perhaps something like:

```
=IIAB=====================================================
/etc/iiab/iiab.env
-IIAB-----------------------------------------------------
bla
bla
=IIAB=====================================================
/opt/iiab/iiab/iiab-network.log -- last 50 lines
-IIAB-----------------------------------------------------
bla
bla
=IIAB=====================================================
/sbin/ip a
-IIAB-----------------------------------------------------
bla
bla
=IIAB=====================================================
/bin/netstat -rn    # routing table
-IIAB-----------------------------------------------------
bla
bla
=IIAB=====================================================
/etc/network/interfaces -- directory contents follow...
-IIAB-----------------------------------------------------
bla
bla
```

2) As above, please do include the full path to all files & commands run if possible?  So we make the diagnostic process inclusive and crystal clear to the OP (original poster) and to community support volunteers learning the ropes.

   So we too stay on top of ever-changing OS quirks during the ever-exhausting-for-everyone-involved debugging/troubleshooting process.  Clean tools take the edge off, and accelerate this painful but necessary work.

   Posting paths that contain 'bin' or 'sbin' or '/usr/bin/iiab-*' is also a great way to succinctly make obvious that what follows is that command's output &mdash; rather than a file's contents.

   Education is an incredibly messy business: but the more we support community support volunteers of every kind, the more IIAB will scale, as they begin to support others using vivid troubleshooting tools like this one.

3) The output should clearly differentiate between an empty file and a non-existent file.  In plain English if possible ("FILE DOES NOT EXIST" or some such, to avoid all ambiguity?)

4) Some might argue that we should include 100 (not 50) last lines of log files, in general in 2019?  I don't know.  50 lines might well be enough for now.  We can easily change this later as pragmatic use/usability patterns become clear.

   And/or soup things up by posting (a) the number of lines contained in each file, or (b) posting the line numbers on the left of file contents &mdash; if the chosen pastebin site could be made to cooperate, etc &mdash; as such, those more advanced things can wait.

5) https://github.com/georgejhunt/iiab/blob/netdiag/scripts/iiab-diagnostics#L72 incorrectly shows "stage 3".  It should say:
```
echo "stage2 = lite; stage5 = desktop SEE github.com/RPi-Distro/pi-gen#stage-anatomy"
```

6) This one's harder but just as real.  Pushing for email addresses and publishing them (without a take-down procedure, etc) can be a serious violation of EU law.  IIAB would be wise not to make itself illegal across most all of Europe.  EU law (and other laws) give people the fundamental right to demand take-down of any PII (personal identifiable information) about them.

   IIAB's community itself expects/deserves privacy to be handled more methodically than this in my opinion, regardless of the law.  Quite aside from the law, it's considered extremely disrespectful to web-publish someone's email address in 2019, no matter whether you live in the EU or not.

   Related: Even publishing a person's public IP address (which their IIAB could very easily do!) starts to become increasingly more and more dangerous.  In 2019, we can no longer ignore core/humanitarian principles of (a) data minimization and (b) personal info being de-identified (or stored in a very different place, e.g. a proper CRM for community relationship mgmt) for ethical, legal, and security reasons.  In short if IIAB develops a reputation for hurting peoples' privacy &mdash; even accidentally &mdash; that will tarnish everything we've done, and permanently.

    Instead: IIAB should take a leadership position in handling privacy in more innovative/subtle/thoughtful ways, in 2019 especially &mdash; whether our community product is used in refugee camps, dictatorial regimes &mdash; or wherever!

7) There are typos like https://github.com/georgejhunt/iiab/blob/netdiag/scripts/iiab-diagnostics#L93 but these visual cleanups will become obvious as this incredibly valuable tool converges & takes root, responsive to IIAB communities' diverse & growing need for understandable & empowering troubleshooting paths...

**Comment by holta** [comment]
@georgejhunt should this PR be closed in favor of PR #1757 ?

**Comment by georgejhunt** [comment]
yes
